Adding basic API rate limiting

robzienert opened the pull request:
I opted for a bucketed approach, rather than a sliding window since it's a lot simpler from a storage perspective. We can avoid horrible, bursty death by making the buckets smaller, which should suffice.

```
rateLimit:
  enabled: true
  learning: true
  redis:
    enabled: true
  capacity: 5
  rateSeconds: 30
  capacityByPrincipal:
    anonymous: 50
```

@spinnaker/netflix-reviewers PTAL

ajordens commented:
Can you document what `capacity`, `rateSeconds` and `principalOverrides` actually are.

`principalOverrides` might be more understandable if named something like `capacityPerPrincipal`? (I think it's a `capacity` specification right)?

What happens if redis connectivity momentarily drops? I think we've revised the health check so that `gate` will no longer automatically drop out of the ELB ... specifically that api requests can still be served even if redis is unavailable. Probably worth preserving that behavior?

ajordens commented:
Might be worth some metrics around throttles and a log message whenever someone does get throttled?

robzienert commented:
There's a log message when people are throttled. I'll add some more
metrics, though, as well as:

1) learning mode (so we can turn on rate limiting without actually 429ing)
2) redis-backed rate overrides.

Should maybe have those out tomorrow. On a plane today.

On Feb 16, 2017 9:06 AM, "Adam Jordens" <notifications@github.com> wrote:

> Might be worth some metrics around throttles and a log message whenever
> someone does get throttled?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/gate/pull/342#issuecomment-280393078>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGoxUyb6zFemKRFpDESAIS_Dic6ayZlks5rdIH_gaJpZM4MBRPJ>
> .
>

ajordens commented:
👍 I missed the log message.

robzienert commented:
Also will make it disable when redis is down. I figure falling back into an
in-memory limiter wouldnt really be too useful.

On Feb 16, 2017 10:02 AM, "Adam Jordens" <notifications@github.com> wrote:

> 👍 I missed the log message.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spinnaker/gate/pull/342#issuecomment-280409157>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGoxXrKPfyX6IpasFIrxf3aQBhHe8QCks5rdI9QgaJpZM4MBRPJ>
> .
>

robzienert commented:
* added docs
* added redis-backed capacityByPrincipal option
* added learning mode
* catching `JedisException` and disabling rate limiter for that request
